src/tasks/mtm.py

```diff
@@ -0,0 +1,297 @@
+# You may obtain a copy of the License at
+# 
+#     https://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Masked Token Modeling (MTM) Task.
+Include Masked Language Modeling (MLM) and Masked Patch Prediction (MPP).
+"""
+
+from typing import Mapping
+
+import dataclasses
+import tensorflow as tf
+
+from official.core import base_task
+from official.core import config_definitions as cfg
+from official.core import task_factory
+from official.modeling import tf_utils
+from official.nlp.data import data_loader_factory
+from official.nlp.modeling import layers
+
+from configs import encoders
+import input_utils
+from modeling import losses
+from configs import mmt
+from modeling import models
+
+
+@dataclasses.dataclass
+class MaskedTMConfig(cfg.TaskConfig):
+  """The model config."""
+  model: mmt.PretrainModelConfig = mmt.PretrainModelConfig()
+
+  scale_loss: bool = False
+  train_data: cfg.DataConfig = cfg.DataConfig()
+  validation_data: cfg.DataConfig = cfg.DataConfig()
+
+
+@task_factory.register_task_cls(MaskedTMConfig)
+class MaskedTMTask(base_task.Task):
+  """Task object for Masked Tokens Modeling (MTM).
+
+  MTM includes Masked Language Modeling (MLM) and Masked Patch Prediction (MPP).
+
+  """
+
+  def _build_encoder(self, encoder_cfg):
+    return encoders.build_encoder(encoder_cfg)
+
+  def build_model(self, params=None):
+    config = params or self.task_config.model
+    encoder_cfg = config.encoder
+    encoder = self._build_encoder(encoder_cfg)
+
+    data_cfg = self.task_config.train_data
+    mpp_output_num_classes = (2 ** data_cfg.output_channel_bits) ** 3
+
+    encoder_inputs = input_utils.create_mmt_encoder_inputs(data_cfg)
+    encoder(**encoder_inputs)
+
+    inputs, labels = input_utils.create_mtm_inputs_and_labels(data_cfg)
+
+    cls_heads = []
+    for cfg in config.cls_heads:
+      cls_heads.append(layers.ClassificationHead(**cfg.as_dict()))
+    cls_labels = input_utils.create_cls_heads_labels(config.cls_heads)
+
+    model = models.MmtPretrainingModel(
+        encoder=encoder,
+        mpp_output_num_classes=mpp_output_num_classes,
+        mlm_activation=tf_utils.get_activation(config.mlm_activation),
+        mlm_initializer=config.mlm_initializer,
+        mpp_activation=tf_utils.get_activation(config.mpp_activation),
+        mpp_initializer=config.mpp_initializer,
+        classification_heads=cls_heads)
+
+    inputs.update(encoder_inputs)
+    model(**inputs)
+
+    return model
+
+  def build_losses(self,
+                   labels,
+                   model_outputs,
+                   metrics,
+                   aux_losses=None) -> tf.Tensor:
+
+    if 'itm_label_weights' in labels:
```

Comment:
The rest of the file talks about MMT, MLM and MPP, but here there is "ITM", is this the same as any of the others, or a new one? if it's a new one, please spell it out in a comment, to clarify.

Reply:
I modified the MTM to PretrainingTask. The pretraining tasks cover MLM, MPP, and ITM.